core/modules/taxonomy/taxonomy.module

```diff
@@ -1852,6 +1854,32 @@ function taxonomy_field_widget_form(&$form, &$form_state, $field, $instance, $la
   return $element;
 }
 
+/**
+ * Implements hook_field_widget_form_alter().
+ */
+function taxonomy_field_widget_form_alter(&$element, &$form_state, $context) {
+  if ($context['field']['type'] == 'taxonomy_term_reference' && in_array($element['#type'], array(
+    'checkboxes',
+    'radios',
+  ))) {
+    $option_extras = array();
+    foreach ($context['field']['settings']['allowed_values'] as $tree) {
+      if ($vocabulary = taxonomy_vocabulary_load($tree['vocabulary'])) {
```

Comment:
I haven't checked everything in depth, but calling `taxonomy_vocabulary_load()` for each option seems a bit excessive. Can't we load the vocabulary once outside the `foreach` loop? Is there any situation where the options loaded are coming from a different vocabulary?

Reply:
`taxonomy_vocabulary_load()` is also called in `taxonomy_allowed_values()` (which is where I cribbed this bit of code from) when the original list of options is built. taxonomy_vocabulary_load() calls `taxonomy_vocabulary_load_multiple()`, which statically caches its result, so I think this call is cheap here.